### ISSUE TITLE
Cannot start - Cast to TV: could not read npm dependencies!

### ISSUE THREAD

**Issue by obi-wil** [Rafostar/gnome-shell-extension-cast-to-tv#124]
1. After a fresh install of this extension and it's node modules from extension prefs, usually a reboot fixes this issue (and by reboot I mean full system reboot - not just restarting gnome-shell).

2. Make sure you have receiver type in extension prefs set correctly to desired device type (e.g. "Chromecast"). If it wasn't try casting again.

3. If extension seems to be working (you have a top panel drop down menu), but still experiencing problems with casting, lets take a look at what happens during file cast. Turn off this extension node service without disabling whole extension (from gnome top bar menu). TV indicator icon should disappear from top bar. Then do the following in terminal:

```
cd ~/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/node_scripts
DEBUG=bridge,chromecast* node server
```

After running, extension menu and indicator should reappear.
Try reproducing your problem and send me output. Stop the terminal process with Ctrl+c buttons.

Output:
```
~/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/node_scripts$ DEBUG=bridge,chromecast* node server
/home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/node_modules/ffprobe-mini/index.js:1
(function (exports, require, module, __filename, __dirname) { const { spawn } = require('child_process');
                                                                    ^

SyntaxError: Unexpected token {
    at exports.runInThisContext (vm.js:53:16)
    at Module._compile (module.js:373:25)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (/home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/node_modules/ffmpeg-extract/index.js:1:79)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
```
Installation went fine but whenever I try to cast anything the tv screen goes black and shows the name of whatever I'm trying to cast but nothing else happens. After a minute or so I get a notification that says: Failed to load media.
Any help would be much appreciated!

**Comment by Rafostar** [comment]
What is your node version?
```
node --version
```

**Comment by obi-wil** [comment]
4.9.1

**Comment by Rafostar** [comment]
> 4.9.1

Too old. Latest stable version is node 12. This extension requires at least one of more recent releases of v8 (that has some of new features backported from v10). Please update your node version.

**Comment by obi-wil** [comment]
I uninstalled and re-installed the extension and now it won't turn on from the panel. Node version is now 14.5.0

**Comment by Rafostar** [comment]
Did you remember to install node modules once again when you reinstalled extension? What happens now when you do:
```
cd ~/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/node_scripts
DEBUG=bridge,chromecast* node server
```

**Comment by obi-wil** [comment]
I installed the node modules again. This is the output of the debug:

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/node_scripts$ DEBUG=bridge,chromecast* node server
  bridge Full playlist: ["/home/wilfredo/Movies/XXXXXXX.S08E06.720p.WEB.x265-MiNX[TGx]/XXXXXX.S08E06.720p.WEB.x265-MiNX.mkv"] +0ms
  bridge New selection contents: {"streamType":"VIDEO","subsPath":"","filePath":"/home/wilfredo/Movies/XXXXXXX.S08E06.720p.WEB.x265-MiNX[TGx]/XXXXXXX.S08E06.720p.WEB.x265-MiNX.mkv"} +0ms
  bridge Processing video file... +1ms
  bridge No reusable subtitles file +1ms
  bridge Extracting video subtitles... +252ms
  bridge Successfully extracted video subtitles +571ms
  bridge File processed successfully +0ms
  chromecast NEW SELECTION +0ms
  chromecast Generated new session id: zkaQ2oEHFZ +7ms
  chromecast Media title: XXXXXXX.S08E06.720p.WEB.x265-MiNX +0ms
  chromecast Setting media: {"contentId":"http://192.168.1.10:4000/cast?session=zkaQ2oEHFZ","contentType":"video/*","streamType":"BUFFERED","metadata":{"metadataType":1,"images":[{"url":""}],"title":"XXXXXXX.S08E06.720p.WEB.x265-MiNX"},"textTrackStyle":{"backgroundColor":"#00000000","foregroundColor":"#FFFFFFFF","edgeType":"OUTLINE","edgeColor":"#000000FF","fontScale":"1.0","fontStyle":"NORMAL","fontFamily":"Droid Sans","fontGenericFamily":"SANS_SERIF","windowType":"NONE"},"tracks":[{"trackId":1,"type":"TEXT","trackContentType":"text/vtt","name":"Subtitles","subtype":"SUBTITLES","trackContentId":"http://192.168.1.10:4000/subswebplayer?session=zkaQ2oEHFZ"}]} +0ms
  chromecast Setting opts: {"autoplay":false,"activeTrackIds":[1],"name":"c21b83a1-1e3b-0de5-b08a-c1964f060a7b.local","ip":null} +0ms
  chromecast Casting... +1ms
  chromecast-controller Player check: inactive +0ms
  chromecast-controller Starting new cast session... +0ms
  chromecast-controller Searching for device: c21b83a1-1e3b-0de5-b08a-c1964f060a7b.local +0ms
  chromecast-controller Found device ip: 192.168.1.216 +378ms
  chromecast-controller Player check: active +1s
  chromecast-controller Could not load media in current session: Load failed +2m
  chromecast-controller Retrying... +0ms
  chromecast-controller Closing client... +0ms
  chromecast-controller Player check: active +0ms
  chromecast-controller Closed client +60ms
  chromecast-controller Connecting to 192.168.1.216 +0ms
  chromecast-controller Player check: active +1s
  chromecast-controller Could not load media in current session: Load failed +2m
  chromecast Could not cast: Load failed +4m
  chromecast Error: Load failed
  chromecast     at /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/node_modules/castv2-client/lib/controllers/media.js:75:23
  chromecast     at fn.onmessage (/home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/node_modules/castv2-client/lib/controllers/request-response.js:27:7)
  chromecast     at fn.emit (events.js:203:15)
  chromecast     at Channel.onmessage (/home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/node_modules/castv2-client/lib/controllers/controller.js:16:10)
  chromecast     at Channel.emit (events.js:198:13)
  chromecast     at Client.onmessage (/home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/node_modules/castv2/lib/channel.js:23:10)
  chromecast     at Client.emit (events.js:203:15)
  chromecast     at PacketStreamWrapper.onpacket (/home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/node_modules/castv2/lib/client.js:81:10)
  chromecast     at PacketStreamWrapper.emit (events.js:198:13)
  chromecast     at TLSSocket.<anonymous> (/home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/node_modules/castv2/lib/packet-stream-wrapper.js:28:16) +0ms
Chromecast: Failed to load media
------------------------------------------------------------------------------------------------------------------------------------------------------------------

**Comment by obi-wil** [comment]
Also tried avi, mp3, jpg with similar outcomes. the screen stays black.

**Comment by Rafostar** [comment]
> Also tried avi, mp3, jpg with similar outcomes. the screen stays black.
Also interesting, casting from the vlc gives me the same outcome. It used to work... But I can cast stuff from chrome and my phone.

If screen stays black try unplugging your Chromecast from AC power outlet and replug after a short while.

The above output is to be expected. Basic Chromecast model does not support x265 videos like you are trying to cast. Try to cast that video with transcode Video+Audio option.

**Comment by obi-wil** [comment]
I have a Chromecast Ultra. It's the Stadia one... I don't think it matters but I know that they're modified: not every chromecast ultra works with Stadia.

**It also doesn't turn on if I exit the debug.**

**Comment by obi-wil** [comment]
Tried the transcode options with several different video formats and still the same. Also tried casting to the android tv built-in chromecast (which is like a conventional one) and still the same.

**Comment by Rafostar** [comment]
> It also doesn't turn on if I exit the debug.

We will get to checking that later. For now we need to determine why you cannot cast at all and for that debug is helpful.

>I have a Chromecast Ultra.

If it is an Ultra then you should be able to cast most of videos without transcoding. If you cannot cast from VLC too, then it probably means that there is something wrong with your firewall configuration in your PC or router.

**Comment by obi-wil** [comment]
Haha, funny. Back to when I could turn on the extension I tried disabling the firewall and nothing happened. Now I disabled the firewall and it works... So I'll be adding some rules. That leave us with the turn on issue. Ideas?

**Comment by Rafostar** [comment]
OK. As I suspected it was indeed your firewall, blocking incoming connections. You can find some basic rules required by Chromecast [here](https://github.com/Rafostar/gnome-shell-extension-cast-to-tv/wiki/Firewall-rules).

> That leave us with the turn on issue. Ideas?

Actually since you had an old `node` version earlier check if you do not have a `nodejs` too. On some distros node was renamed to nodejs to avoid naming conflicts. The extension first tries to use nodejs and then fallbacks to node. The below command should not give you any output:
```
nodejs --version
```
If you have `nodejs`, remove it so only your `node` you installed recently remains.

Also did you try to reboot after you reinstalled the extension? GNOME extensions sometimes do not work until you reboot/logout after fresh install.

If you check above two solutions to no avail. Try clicking turn on and then check your journalctl output and try to find any cast related messages:
```
journalctl -b | grep -i cast
```

**Comment by obi-wil** [comment]
Actually since you had an old node version earlier check if you do not have a nodejs too. On some distros node was renamed to nodejs to avoid naming conflicts. The extension first tries to use nodejs and then fallbacks to node. The below command should not give you any output:

The thing was that I had an old node version and the newer nodejs version from the focal distro... Bit messy... I uninstalled everything, installed the 14.5.0 version from repo and then couldn't install nodejs and it was giving me some weird behaviors. Then I just installed the distro version so both node and nodejs are 10.19.0

I rebooted after reinstalling the extension.

This is the journalctl output:

Jul 07 14:20:37 wilfredo-Racer-4W org.gnome.Shell.Extensions[33203]: #2   55618f26d7b8 i   /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/soup.js:421 (18a537dfbda8 @ 219)
Jul 07 14:20:37 wilfredo-Racer-4W gjs[33203]: Object .Gjs_CastNotebook (0x55618fa0e290), has been already deallocated — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
Jul 07 14:49:37 wilfredo-Racer-4W org.gnome.Shell.Extensions[33203]: #0   55618f26d910 i   /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/prefs.js:219 (18a537df2718 @ 110)
Jul 07 14:49:37 wilfredo-Racer-4W org.gnome.Shell.Extensions[33203]: #1   55618f26d878 i   /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/prefs.js:1146 (18a537df7100 @ 161)
Jul 07 14:49:37 wilfredo-Racer-4W org.gnome.Shell.Extensions[33203]: #2   55618f26d7b8 i   /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/soup.js:421 (18a537dfbda8 @ 219)
Jul 07 14:49:37 wilfredo-Racer-4W org.gnome.Shell.Extensions[33203]: #0   55618f26d910 i   /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/prefs.js:220 (18a537df2718 @ 130)
Jul 07 14:49:37 wilfredo-Racer-4W org.gnome.Shell.Extensions[33203]: #1   55618f26d878 i   /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/prefs.js:1146 (18a537df7100 @ 161)
Jul 07 14:49:37 wilfredo-Racer-4W org.gnome.Shell.Extensions[33203]: #2   55618f26d7b8 i   /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/soup.js:421 (18a537dfbda8 @ 219)
Jul 07 14:49:43 wilfredo-Racer-4W gjs[39094]: JS LOG: Cast to TV: could not read npm dependencies!

**Comment by Rafostar** [comment]
```
Jul 07 14:20:37 wilfredo-Racer-4W org.gnome.Shell.Extensions[33203]: #2 55618f26d7b8 i /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/soup.js:421 (18a537dfbda8 @ 219)
Jul 07 14:20:37 wilfredo-Racer-4W gjs[33203]: Object .Gjs_CastNotebook (0x55618fa0e290), has been already deallocated — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
Jul 07 14:49:37 wilfredo-Racer-4W org.gnome.Shell.Extensions[33203]: #0 55618f26d910 i /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/prefs.js:219 (18a537df2718 @ 110)
Jul 07 14:49:37 wilfredo-Racer-4W org.gnome.Shell.Extensions[33203]: #1 55618f26d878 i /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/prefs.js:1146 (18a537df7100 @ 161)
Jul 07 14:49:37 wilfredo-Racer-4W org.gnome.Shell.Extensions[33203]: #2 55618f26d7b8 i /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/soup.js:421 (18a537dfbda8 @ 219)
Jul 07 14:49:37 wilfredo-Racer-4W org.gnome.Shell.Extensions[33203]: #0 55618f26d910 i /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/prefs.js:220 (18a537df2718 @ 130)
Jul 07 14:49:37 wilfredo-Racer-4W org.gnome.Shell.Extensions[33203]: #1 55618f26d878 i /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/prefs.js:1146 (18a537df7100 @ 161)
Jul 07 14:49:37 wilfredo-Racer-4W org.gnome.Shell.Extensions[33203]: #2 55618f26d7b8 i /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/soup.js:421 (18a537dfbda8 @ 219)
```

Those are related to prefs (extension settings) window only on GNOME 3.36. Should not affect turn on/casting. I am aware of those and will try to fix them when I have more free time. For now ignore...

```
Jul 07 14:49:43 wilfredo-Racer-4W gjs[39094]: JS LOG: Cast to TV: could not read npm dependencies!
```

This is the reason why turn on does not work for you. It looks like for some reason the extension cannot access (and read content of) `package.json` file inside the extension folder. It should be found in below path:
```
~/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/package.json
```

I do not know how you reinstalled the extension, but it seems that file was accessible from extension before reinstall, but now extension cannot read it.

**Comment by obi-wil** [comment]
Okay. Reinstalled the extension and got a notification: Missing npm module body-parser. Installed it and this is what I got upon installation:

npm install body-parser
npm WARN saveError ENOENT: no such file or directory, open '/home/wilfredo/package.json'
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN enoent ENOENT: no such file or directory, open '/home/wilfredo/package.json'
npm WARN wilfredo No description
npm WARN wilfredo No repository field.
npm WARN wilfredo No README data
npm WARN wilfredo No license field.

+ body-parser@1.19.0
added 22 packages from 17 contributors and audited 22 packages in 1.94s
found 0 vulnerabilities

From here I'm lost...

**Comment by Rafostar** [comment]
> npm install body-parser

Please do not do that. The extension already has a preconfigured package.json and it "knows" what package at what version is needed. If you do it manually you will mess up the package.json and package-lock configs!!!!

1. Remove current broken install:
```
rm -rf /home/wilfredo/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com
```
2. logout and login back
3. install and enable extension from GNOME Extensions website
4. open prefs (cast settings) -> go to modules tab -> Click "Install NPM modules" button.

You should be able to turn it on now.

**Comment by obi-wil** [comment]
Did all of it and it persists:

Jul 07 16:39:14 wilfredo-Racer-4W gjs[29855]: JS LOG: Cast to TV: npm modules not installed!
Jul 07 16:40:00 wilfredo-Racer-4W gjs[30908]: JS LOG: Cast to TV: could not read npm dependencies!
Jul 07 16:40:50 wilfredo-Racer-4W gjs[31001]: JS LOG: Cast to TV: could not read npm dependencies!
Jul 07 16:40:57 wilfredo-Racer-4W gjs[31025]: JS LOG: Cast to TV: could not read npm dependencies!

**Comment by Rafostar** [comment]
Can you send me the content of your `~/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/package.json` file?

**Comment by obi-wil** [comment]
{
  "name": "cast-to-tv",
  "description": "Cast files to Chromecast, web browser or media player app over local network.",
  "main": "node_scripts/server.js",
  "author": "Rafostar",
  "license": "GPL-2.0",
  "dependencies": {
    "body-parser": "^1.19.0",
    "bowser": "^2.9.0",
    "chromecast-controller": "^0.3.1",
    "debug": "^4.1.1",
    "express": "^4.17.1",
    "ffmpeg-extract": "^1.1.2",
    "gettext-parser": "^3.1.0",
    "internal-ip": "^4.3.0",
    "multicast-scanner": "^2.2.1",
    "node-gettext": "^2.0.0",
    "plyr": "3.5.3",
    "socket.io": "^2.3.0",
    "ws": "^7.2.1"
  },
  "scripts": {
    "install": "sh node_scripts/utils/install.js"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Rafostar/gnome-shell-extension-cast-to-tv.git"
  },
  "keywords": [
    "gnome-shell-extension",
    "chromecast",
    "media",
    "video",
    "stream",
    "cast"
  ],
  "bugs": {
    "url": "https://github.com/Rafostar/gnome-shell-extension-cast-to-tv/issues"
  },
  "homepage": "https://github.com/Rafostar/gnome-shell-extension-cast-to-tv",
  "devDependencies": {
    "bufferutil": "^4.0.1",
    "utf-8-validate": "^5.0.2"
  }
}

**Comment by Rafostar** [comment]
OK. File exists and content is alright. I would like to see this file owner and permissions. Can you run `ls -la` inside extension folder?

**Comment by obi-wil** [comment]
Solved.
I found out that the links addon folder was still around. I'm sure I uninstalled both desktop and links addons but... Links was there in the extensions folder. I removed it and it now turns on. Does that make sense?

**Comment by Rafostar** [comment]
> I found out that the links addon folder was still around. I'm sure I uninstalled both desktop and links addons but... Links was there in the extensions folder. I removed it and it now turns on. Does that make sense?

Yes. It does make a lot of sense. The extension reads the package.json durring startup, but if it detects addons are installed it also reads the package.json files inside the addons folders. You removed the addon but the folder was left so extension tried to read package.json inside Links Add-on folder and failing.

Since you never managed to make the extension work in the first place I thought that you havent tried to install any addons yet. So the broken package.json was in addon folder and not in main extension folder.

I should probably learn from this and improve the logs to also log where was the file it could not read instead of only filename.

**Comment by obi-wil** [comment]
Sure. It's funny how simple stuff are always so easy to overview. I spent the afternoon uninstalling and installing node lol. It also didn't occurred to me to check the folder and the files. Also because I was moving stuff around all at the same time. Not so great for debugging...

I was just checking it out now. It's amazing! The desktop doesn't quite get there yet. Too much delay, especially with the cursor drawing. But quite stable with nvenc at 10 Mbps 60 fps.

Thanks a lot for your help. You're truly dedicated. This would be a better world if 1/1000 of the tech support guys would have the patience. I'll drop some money into the project when I'm not so broke (damn covid).

**Comment by Rafostar** [comment]
Thank you for cooperation too and also thanks for trying this extension.

Lets leave this one issue open for now, so I won't forget to improve the startup logs to be more precise in next release.

**Comment by obi-wil** [comment]
Good. Let me know if you need help testing or something. Bis bald!

**Comment by Rafostar** [comment]
There were added more checks during startup some time ago. The most recent commits should add the path of folder where "package.json" was unreadable, so hopefully similar issue won't happen anymore. The "Extensions" app errors from journal should now be fixed too. We cannot really check for `nodejs` version from extension system in a efficient way, hopefully not many people should have an old version like you did. I updated readme to add info about minimal required version.

I cannot guarantee that latest git code works as reliably as last stable release (there were a lot of changes recently - some breaking), so you might be better off staying on the installation you already have if it works for you until next "stable" release.
